tket2-py/tket2/circuit/build.py

```diff
@@ -1,137 +1,189 @@
         and validate."""
-        return self.dfg.finish(self.qbs)
+        return load_hugr(self.hugr)
 
 
 def from_coms(*args: Command) -> Tk2Circuit:
-    """Build a circuit from a sequence of commands, assuming only qubit outputs."""
-    commands = []
+    """Build a circuit from a sequence of commands, assuming
+    only qubits are referred to by index."""
+    commands: list[Command] = []
     n_qb = 0
     # traverses commands twice which isn't great
     for arg in args:
-        max_qb = max(arg.qubits()) + 1
+        max_qb = max(i for i in arg.incoming if isinstance(i, int)) + 1
         n_qb = max(n_qb, max_qb)
         commands.append(arg)
 
-    build = CircBuild(n_qb)
-    build.extend(commands)
+    build = CircBuild.with_nqb(n_qb)
+    build.extend(*commands)
+    build.set_tracked_outputs()
     return build.finish()
 
 
-# Some common operations
+def load_hugr(h: Hugr) -> Tk2Circuit:
+    return Tk2Circuit.from_hugr_json(h.to_serial().model_dump_json())
+
+
+def load_custom(serialized: bytes) -> ops.Custom:
+    import hugr.serialization.ops as sops
+    import json
+
+    dct = json.loads(serialized)
+    dct["parent"] = -1
+    return sops.CustomOp(**dct).deserialize()
+
+
+def id_circ(n_qb: int) -> Tk2Circuit:
+    b = CircBuild.with_nqb(n_qb)
+    b.set_tracked_outputs()
+    return b.finish()
+
+
+@dataclass(frozen=True)
+class QuantumOps(ops.Custom):
+    extension: tys.ExtensionId = "quantum.tket2"
+
+
+_OneQbSig = tys.FunctionType.endo([tys.Qubit])
+
+
+@dataclass(frozen=True)
+class OneQbGate(QuantumOps):
+    op_name: str  # type: ignore[misc] # no-default fields follows one with a default
+    num_out: int = 1
+    signature: tys.FunctionType = _OneQbSig
+
+    def __call__(self, q: ComWire) -> Command:
+        return super().__call__(q)
+
+
+H = OneQbGate("H")
+PauliX = OneQbGate("X")
+PauliY = OneQbGate("Y")
+PauliZ = OneQbGate("Z")
 
-# Define some "Commands" for pure quantum gates (n qubits in and n qubits out)
+_TwoQbSig = tys.FunctionType.endo([tys.Qubit] * 2)
 
 
 @dataclass(frozen=True)
-class H(Command):
-    qubit: int
-    gate_name = "H"
-    n_qb = 1
+class TwoQbGate(QuantumOps):
+    op_name: str  # type: ignore[misc] # no-default fields follows one with a default
+    num_out: int = 2
+    signature: tys.FunctionType = _TwoQbSig
 
-    def qubits(self) -> list[int]:
-        return [self.qubit]
+    def __call__(self, q0: ComWire, q1: ComWire) -> Command:
+        return super().__call__(q0, q1)
+
+
+CX = TwoQbGate("CX")
+
+_MeasSig = tys.FunctionType([tys.Qubit], [tys.Qubit, tys.Bool])
+
+
+@dataclass(frozen=True)
+class MeasureDef(QuantumOps):
+    op_name: str = "Measure"
+    num_out: int = 2
+    signature: tys.FunctionType = _MeasSig
+
+    def __call__(self, q: ComWire) -> Command:
+        return super().__call__(q)
+
+
+Measure = MeasureDef()
+
+
+FLOAT_T = tys.Opaque(
+    extension="arithmetic.float.types",
+    id="float64",
+    args=[],
+    bound=tys.TypeBound.Copyable,
+)
```

Comment:
This should be defined in `hugr-py`. Open an issue there?

Reply:
will do a PR adding some prelude types

Reply:
https://github.com/CQCL/hugr/pull/1288